src/snowcli/cli/common/cli_global_context.py

```diff
@@ -0,0 +1,146 @@
+            if override == "connection" or override == "temporary_connection":
+                continue
+            override_value = getattr(self, override)
+            if override_value is not None:
+                params[override] = override_value
+        return params
+
+    def build_connection(self):
+        return connect_to_snowflake(
+            temporary_connection=self.temporary_connection,
+            connection_name=self.connection_name,
+            **self._resolve_connection_params()
+        )
+
+
+DEFAULT_ENABLE_TRACEBACKS = True
+DEFAULT_OUTPUT_FORMAT = OutputFormat.TABLE
+DEFAULT_VERBOSE = False
+DEFAULT_EXPERIMENTAL = False
+
+
+class _GlobalContextManager:
+    _connection_details = ConnectionDetails()
+    _cached_connection: Optional[SnowflakeConnection] = None
+
+    enable_tracebacks = DEFAULT_ENABLE_TRACEBACKS
+    output_format = DEFAULT_OUTPUT_FORMAT
+    verbose = DEFAULT_VERBOSE
+    experimental = DEFAULT_EXPERIMENTAL
+
+    def set_default_values(self):
+        self.connection_details = ConnectionDetails()
+        self.enable_tracebacks = DEFAULT_ENABLE_TRACEBACKS
+        self.output_format = DEFAULT_OUTPUT_FORMAT
+        self.verbose = DEFAULT_VERBOSE
+        self.experimental = DEFAULT_EXPERIMENTAL
+
+    @property
+    def connection_details(self) -> ConnectionDetails:
+        return self._connection_details
+
+    @connection_details.setter
+    def connection_details(self, connection_details: ConnectionDetails):
+        self._connection_details = connection_details
+        self._cached_connection = None
+
+    @property
+    def cacheable_connection(self) -> SnowflakeConnection:
+        if not self._cached_connection:
+            self._cached_connection = self.connection_details.build_connection()
+        return self._cached_connection
+
+    def update_global_context_option(self, param_name: str, value: Union[bool, str]):
+        setattr(self, param_name, value)
+
+    def update_global_connection_detail(self, param_name: str, value: str):
+        setattr(self._connection_details, param_name, value)
+        self._cached_connection = None
+
+
+class _GlobalContextAccess:
+    def __init__(self, manager: _GlobalContextManager):
+        self._manager = manager
+
+    @property
+    def connection_details(self) -> ConnectionDetails:
+        return self._manager.connection_details
+
+    @property
+    def cacheable_connection(self) -> SnowflakeConnection:
+        return self._manager.cacheable_connection
+
+    @property
+    def enable_tracebacks(self) -> bool:
+        return self._manager.enable_tracebacks
+
+    @property
+    def output_format(self) -> OutputFormat:
+        return self._manager.output_format
+
+    @property
+    def verbose(self) -> bool:
+        return self._manager.verbose
+
+    @property
+    def experimental(self) -> bool:
+        return self._manager.experimental
+
+
+global_context_manager: _GlobalContextManager = _GlobalContextManager()
```

Comment:
Do we use it somewhere outside this module?

Reply:
Currently only in `conftest.py` files in unit and integration tests.

But I think it should be a public field because you may want to change the global context from the code in other places than callbacks to global options.

`global_context_manager` is separated from `global_context`, so if you use `global_context_manager` which has "write" access, you do it on purpose. If you just need "read" access, then you import `global_context`. IMO both should be public and we should suggest to use `global_context` and only mention that you have a possibility to change the context via `global_context_manager`.

Reply:
Sounds good to me. The only concern that I still have is the naming, python has already concept of "context manager" 😢  not sure if we can do something about it

Reply:
`snowcli_context_manager`? 😄

Reply:
Names changed to `cli_context` and `cli_context_manager`